### ISSUE TITLE
[Feature]: Support HNSW SQ/PQ

### ISSUE THREAD

**Issue by xiaofan-luan** [milvus-io/milvus#23232]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

SQ8 and PQ are widely used in ANN search. If you want to understand more about quantization, Faiss is probably one of the best code bases to explore.

HNSW is the fastest index in the open source world, so why not make it work together with SQ and PQ to accelerate it further?

Let me know if anyone is interested and we can offer more help on it

**Comment by noble-8** [comment]
Hi @xiaofan-luan
I am interested in contributing!
and would like to know how to help! let me know how to get started.

**Comment by xiaofan-luan** [comment]
cool man! I though @liliu-z could offer you some help

**Comment by xiaofan-luan** [comment]
you have any experience on cpp and some any idea about HSNW algorithm yet?

**Comment by noble-8** [comment]
Hi 
I had taken a cpp course in college, I have primarily worked as a Java developer ~3 ish years, so i feel that I can onboard quickly. I feel that I am comfortable working on cpp.
I am really new to this algorithm but getting up to speed. I am going through this documentation here:https://www.pinecone.io/learn/hnsw/
feel free to point me to other resources.
I hope this is not a dealbreaker!

**Comment by xiaofan-luan** [comment]
1. you can start from read code from HNSW in knowhere https://github.com/milvus-io/knowhere/tree/c05c8767f43eaa855c13654804d0bea9cc42c7de/src/index/hnsw
2. Once you understand hnsw, then the next step would be under stand how to do PQ, SQ. Faiss document will give your general ideas and milvus has all faiss code you can utilize https://github.com/facebookresearch/faiss/wiki.
3. Add index parameters for milvus to support PQ, SQ, which will be a trivial task

**Comment by liliu-z** [comment]
Hi @noble-8 this is on our roadmap and please feel free to make a PR for https://github.com/milvus-io/knowhere . I suggest we can start from SQ8 which is easier to implement. Plz feel free to open another issue in Knowhere for further detailed communication.

**Comment by noble-8** [comment]
Sounds good. Will do!

**Comment by xiaofan-luan** [comment]
@noble-8 
any progress on it?

**Comment by noble-8** [comment]
i could not make any progress. i shall try again, however feel free to reassign this if i do not make a commit

**Comment by xiaofan-luan** [comment]
> i could not make any progress. i shall try again, however feel free to reassign this if i do not make a commit

Sure, still thanks for the interest!
I would also like to help if you are intersted

**Comment by LaPetiteSouris** [comment]
I just saw that https://github.com/milvus-io/knowhere is now archived. Wondering if this issue is still open ?

**Comment by xiaofan-luan** [comment]
> https://github.com/milvus-io/knowhere

It has been archived and moved to https://github.com/zilliztech/knowhere, sorry for the misunderstanding.

You are correct my man. we want to add quantization support for HNSW index and integrate with Milvus

**Comment by LaPetiteSouris** [comment]
Thanks. How urgent do you folks need this ? My C++ is rusty 😭 so it may take a while ( I have CoPilot so that helps 😭  ). 

But I love this challenge.

**Comment by jiaoew1991** [comment]
@xiaofan-luan I think the issue is not easy for beginners, it needs lots of knowledge 😅

**Comment by xiaofan-luan** [comment]
> @xiaofan-luan I think the issue is not easy for beginners, it needs lots of knowledge 😅

Agreed you might be correct.

For SQ might be ok?

**Comment by xiaofan-luan** [comment]
But true it has to be fully understand milvus

**Comment by xiaofan-luan** [comment]
remove the good first issue

**Comment by zaobao** [comment]
I sound that HNSW-SQ8 has been available on Ziili Cloud. Is that true?

**Comment by xiaofan-luan** [comment]
Zilliz cloud don't use HSNW. we have an internal index named Cardinal~

**Comment by Monster880** [comment]
whether milvus can use hnsw pq index now ?

**Comment by xiaofan-luan** [comment]
@liliu-z 
do we have plan to support hnsw pq and sq index?

**Comment by Monster880** [comment]
@xiaofan-luan  Can you provide some guidance on where modifications are needed to support HNSW PQ index??

**Comment by xiaofan-luan** [comment]
NP, I thought Li @liliu-z can help on that.

**Comment by Monster880** [comment]
@liliu-z  @xiaofan-luan emmm.... where is liliu-z

**Comment by liliu-z** [comment]
> @liliu-z @xiaofan-luan emmm.... where is liliu-z

Sure, there are two ways to support HNSW + Quantization:
1. Make it a new index type for Milvus
2. Treat it as HNSW with a special config.

We are adopting the first way. So the work including:
1. Add the quantization support in algorithm side and expose it as a new index type. Code should be in [Knowhere](https://github.com/zilliztech/Knowhere)
2. Let Milvus know this new Index Type

Here is an [example PR](https://github.com/zilliztech/knowhere/pull/311/files) for the first step. It support SQ8 for HNSW in Knowhere side.

**Comment by Monster880** [comment]
@liliu-z  it seems HNSW_PQ is not using faiss but using  hnswlib after quantization...

**Comment by xiaofan-luan** [comment]
Now we prefer to use hnswlib rather than faiss for hnsw, so we need to backport pq and sq feature

**Comment by Monster880** [comment]
@liliu-z  @xiaofan-luan  I find it is too hard for me to support pq feature for the backport to hnswlib but the hnsw pq  is very beneficial and important for me so that I hope milvus can support hnsw pq index as soon as possible..

**Comment by Monster880** [comment]
@xiaofan-luan I think maybe it can temporary support hnsw pq index by faiss

**Comment by liliu-z** [comment]
> @xiaofan-luan I think maybe it can temporarily support hnsw pq index by faiss

Yes, we are discussing with faiss team about the possibility to switch to faiss' HNSW. There are still some gaps like performance, features and APIs. And we will support pq in hnswlib if we finally decide not to go with faiss.

Will keep this post updated if any progress.

**Comment by Monster880** [comment]
@liliu-z I get it and want to know when can we use hnsw pq index by milvus

**Comment by xiaofan-luan** [comment]
@alexanderguzhva 
please help on this. 
we want make sure faiss HNSW has similar performance and exact same functionality with current knowhere implementation

**Comment by Monster880** [comment]
@xiaofan-luan @liliu-z Since knowHere already supports hnsw_sq milvus, when will milvus support it..

**Comment by alexanderguzhva** [comment]
@xiaofan-luan I'm in the middle of deprecating hwnslib in favor of faiss already, work has been in the progress for some time

**Comment by Monster880** [comment]
@xiaofan-luan Actually, I want to know when will milvus support hnsw_sq index, since knowhere already supports hnsw_sq index..

**Comment by liliu-z** [comment]
> @xiaofan-luan Actually, I want to know when will milvus support hnsw_sq index, since knowhere already supports hnsw_sq index..

It has not been fully tested yet, we can try to support it out as a beta function in the next release, which is 1-2 weeks from now. What do you think @xiaofan-luan

**Comment by xiaofan-luan** [comment]
@liliu-z please make sure knowhere side is ready

@tedxu could you assign someone to support HSNW PQ/SQ and did some test

**Comment by Monster880** [comment]
Acutally, offline data pipeline can load index fasterly and avoid train index process. Is there some solution for me to train index offlinely and load index at container(stand-alone) startup. @xiaofan-luan

**Comment by xiaofan-luan** [comment]
this might be our goal to do so. 

using milvus with more offline index node should help.

if you already index by your self, why not simply serve it with faiss or hnsw?
